app/processImage.js

```diff
@@ -78,24 +79,80 @@ function applySmartCrop(image, cropSize, callback) {
+            applySmartCrop(image, cropSize, callback)
+        } else {
+            callback(null, image, {});
+        }
+    }
 }
 
 function applyCoverResize(image, cropSize, callback) {
-    sharp(image)
-        .resize(cropSize.width, cropSize.height)
-        .max()
-        .toBuffer(callback)
-}
 
-function rawImage(image, cropSize, callback) {
-    if (cropSize.width || cropSize.height) {
-        callback({}); // raise error if height or width is provided with rawType
-        return ;
+    if (cropSize.widthPlus || cropSize.heightPlus) {
+        sharp(image)
+            .resize(cropSize.width, cropSize.height)
+            .max()
+            .toBuffer(callback)
+    } else {
+        if (cropSize.width < cropSize.originalWidth || cropSize.height < cropSize.originalHeight) {
+            sharp(image)
+                .resize(cropSize.width, cropSize.height)
+                .max()
+                .toBuffer(callback)
+        } else {
+            callback(null, image, {});
+        }
     }
-    callback(null, image, {});
+}
+
+function applyBlurEffect(image, cropSize, callback) {
+    let blurImg = sharp(image);
+    let overlayImg = sharp(image);
+
+    blurImg
+        .metadata()
+        .then(function(metadata) {
+            if (cropSize.widthPlus && cropSize.heightPlus) {
+                return applySmartCrop(image, cropSize, callback)
+            } else if (cropSize.widthPlus && !cropSize.heightPlus){
+                cropSize.height = Math.round(Math.min(cropSize.height, metadata.height * (cropSize.width / metadata.width), metadata.height));
+
+                if (metadata.width >= cropSize.width) {
+                    return applySmartCrop(image, cropSize, callback)
+                } else if (metadata.width < cropSize.width && metadata.height > cropSize.height) {
+                    overlayImg
+                        .resize(metadata.width, cropSize.height)
+                }
+            } else if (!cropSize.widthPlus && cropSize.heightPlus) {
+                cropSize.width = Math.round(Math.min(cropSize.width, metadata.width * (cropSize.width / metadata.height), metadata.width));
+
+                if (metadata.height >= cropSize.height) {
+                    return applySmartCrop(image, cropSize, callback)
+                } else if (metadata.height < cropSize.height && metadata.width > cropSize.width) {
+                    overlayImg
+                        .resize(cropSize.width, metadata.height)
+                }
+            } else {
+                if (cropSize.width < cropSize.originalWidth || cropSize.height < cropSize.originalHeight) {
+                    overlayImg
+                        .resize(cropSize.width, cropSize.height)
+                        .max()
+                }
+            }
+
+            overlayImg
+                .toBuffer(function(err, data, info) {
+                    blurImg
```

Comment:
this doesn't handle the err condition. 
we would need check `if ( err == null) { //success  } else { //error }`